File: Dynamic-Programming/1547. Minimum Cost to Cut a Stick/minimumCosttoCutaStick.js
/**
 * @param {number} n
 * @param {number[]} cuts
 * @return {number}
 */
var minCost = function (n, cuts) {
    let memo = new Map();
    return dp(0, n);

    function dp(start, end) {
        let key = start + "-" + end;

        if (memo.has(key)) {
            return memo.get(key)
        }

        let canCut = false;
        let result = Infinity;

        for (const cut of cuts) {
            if (cut > start && cut < end) {
                canCut = true;
                result = Math.min(result, dp(start, cut) + dp(cut, end) + (end - start));
            }
        }

        if (canCut === false) {
            return 0;
        }

        memo.set(key, result);
        return result;
    }
};

File: Dynamic-Programming/1547. Minimum Cost to Cut a Stick/minimumCosttoCutaStick.py
class Solution:
    # n^3
    def minCost(self, n: int, cuts: List[int]) -> int:
        def dp (start, end):
            if (start, end) in memo:
                return memo[(start, end)]
        
            # set ans to infinite           
            ans = sys.maxsize
            # base case
            canCut = False
            
            for cut in cuts:
                if start < cut < end:
                    canCut = True
                    # cost: end - start
                    ans = min(ans, dp(start, cut) + dp(cut, end) + end - start)
            
            if not canCut:
                # if to save the result, 
                # so next time it will return directly from memo
                # use ans = 0 
                return 0
            
            # dict[key]
            memo[(start, end)] = ans

            return ans
            
        memo = {}

        return dp(0, n)